File: 0040-combination-sum-ii/0040-combination-sum-ii.py
class Solution:
    def combinationSum2(self, candidates: List[int], target: int) -> List[List[int]]:
        # first sort the array to prevent duplicate
        candidates.sort()

        ans = []

        def backtrack(i, remain, subset):
            if remain == 0:
                ans.append(subset.copy())
                return
            
            if i == len(candidates) or remain < 0:
                return
            
            subset.append(candidates[i])
            backtrack(i+1, remain-candidates[i], subset)
            subset.pop()

            i += 1
            while i < len(candidates) and candidates[i] == candidates[i-1]:
                i += 1
            backtrack(i, remain, subset)
        
        backtrack(0, target, [])
        return ans

File: 0040-combination-sum-ii/0040-combination-sum-ii.cpp
class Solution {
public:
    void backtrack(vector<int>& candidates, int i, int remain, vector<int>& subset, vector<vector<int>>& ans) {
        if (remain == 0) {
            ans.push_back(subset);
            return;
        }

        if (i == candidates.size() || remain < 0) {
            return;
        }

        subset.push_back(candidates[i]);
        backtrack(candidates, i+1, remain-candidates[i], subset, ans);
        subset.pop_back();
        i++;
        while (i < candidates.size() && candidates[i] == candidates[i-1]) {
            i++;
        }
        backtrack(candidates, i, remain, subset, ans);
    }

    vector<vector<int>> combinationSum2(vector<int>& candidates, int target) {
        sort(candidates.begin(), candidates.end());

        vector<vector<int>> ans;
        vector<int> subset;

        backtrack(candidates, 0, target, subset, ans);
        return ans;
    }
};